engine/src/main/java/org/pentaho/di/job/entries/writetofile/JobEntryWriteToFile.java

```diff
@@ -0,0 +1,315 @@
+    List<SlaveServer> slaveServers ) throws KettleException {
+    try {
+      filename = rep.getJobEntryAttributeString( id_jobentry, "filename" );
+      createParentFolder = rep.getJobEntryAttributeBoolean( id_jobentry, "createParentFolder" );
+      appendFile = rep.getJobEntryAttributeBoolean( id_jobentry, "appendFile" );
+      content = rep.getJobEntryAttributeString( id_jobentry, "content" );
+      encoding = rep.getJobEntryAttributeString( id_jobentry, "encoding" );
+    } catch ( KettleException dbe ) {
+      throw new KettleException(
+        "Unable to load job entry of type 'create file' from the repository for id_jobentry=" + id_jobentry, dbe );
+    }
+  }
+
+  public void saveRep( Repository rep, IMetaStore metaStore, ObjectId id_job ) throws KettleException {
+    try {
+      rep.saveJobEntryAttribute( id_job, getObjectId(), "filename", filename );
+      rep.saveJobEntryAttribute( id_job, getObjectId(), "createParentFolder", createParentFolder );
+      rep.saveJobEntryAttribute( id_job, getObjectId(), "appendFile", appendFile );
+      rep.saveJobEntryAttribute( id_job, getObjectId(), "content", content );
+      rep.saveJobEntryAttribute( id_job, getObjectId(), "encoding", encoding );
+    } catch ( KettleDatabaseException dbe ) {
+      throw new KettleException( "Unable to save job entry of type 'create file' to the repository for id_job="
+        + id_job, dbe );
+    }
+  }
+
+  public void setFilename( String filename ) {
+    this.filename = filename;
+  }
+
+  public String getFilename() {
+    return filename;
+  }
+
+  public void setContent( String content ) {
+    this.content = content;
+  }
+
+  public String getContent() {
+    return content;
+  }
+
+  public void setEncoding( String encoding ) {
+    this.encoding = encoding;
+  }
+
+  public String getEncoding() {
+    return encoding;
+  }
+
+  public String getRealFilename() {
+    return environmentSubstitute( getFilename() );
+  }
+
+  public Result execute( Result previousResult, int nr ) {
+    Result result = previousResult;
+    result.setResult( false );
+    result.setNrErrors( 1 );
+
+    String realFilename = getRealFilename();
+    if ( !Utils.isEmpty( realFilename ) ) {
+
+      //Set Embedded NamedCluter MetatStore Provider Key so that it can be passed to VFS
+      if ( parentJobMeta.getNamedClusterEmbedManager() != null ) {
+        parentJobMeta.getNamedClusterEmbedManager()
+          .passEmbeddedMetastoreKey( this, parentJobMeta.getEmbeddedMetastoreProviderKey() );
+      }
+
+      String content = environmentSubstitute( getContent() );
+      String encoding = environmentSubstitute( getEncoding() );
+
+      OutputStreamWriter osw = null;
+      OutputStream os = null;
+      try {
+
+        // Create parent folder if needed
+        createParentFolder( realFilename );
+
+        // Create / open file for writing
+        os = KettleVFS.getOutputStream( realFilename, this, isAppendFile() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *os*. The resource is referenced at lines 193, 194, 207, 213, 215, 221, 222, 223, 225, 230, 232. The resource is closed at lines 226, 233. There are other execution paths that do not contain closure statements, for example, when an exception is thrown by Writer.write, OutputStream.flush, OutputStreamWriter.flush.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).